util/activation_memory_profiler.py

```diff
@@ -0,0 +1,134 @@
+import json
+from dataclasses import dataclass, field
+from typing import List
+
+import executorch.exir.memory as memory
+import torch
+from executorch.exir import ExecutorchProgramManager
+from executorch.exir.memory_planning import get_node_tensor_specs
+from executorch.exir.print_program import _stacktrace_to_framelist
+from executorch.exir.tensor import num_bytes_from_shape_and_dtype, TensorSpec
+from torch._export import ExportedProgram
+
+
+@dataclass
+class Allocation:
+    name: str
+    op_name: str
+    memory_id: int
+    size_bytes: int
+    fqn: str
+    file_and_line_num: str
+
+
+@dataclass
+class MemoryTimeline:
+    allocations: List[Allocation] = field(default_factory=list)
+
+
+def _get_module_hierarchy(node: torch.fx.Node) -> str:
+    """
+    Get the module hierarchy of the given node.
+    """
+    module_stack = node.meta.get("nn_module_stack")
+    if module_stack is not None:
+        module_values_list = list(module_stack.values())
+        return module_values_list[-1][0]
+    return ""
+
+
+def create_tensor_allocation_info(graph: torch.fx.Graph) -> List[MemoryTimeline]:
+    """
+    Create a list of TensorAllocationInfo from the given nodes.
+    The returned list is sorted by the memory id of each allocation info.
+    """
+    nodes = graph.nodes
+    memory_timeline = [None] * len(nodes)
+    for i, node in enumerate(nodes):
+        tensor_specs = get_node_tensor_specs(node)
+        if tensor_specs is None:
+            continue
+        if node.target == memory.alloc:
+            continue
+        for tensor_spec in tensor_specs:
+            if tensor_spec is None or tensor_spec.mem_id is None or tensor_spec.const:
+                continue
+            start, end = tensor_spec.lifetime
+            size = num_bytes_from_shape_and_dtype(tensor_spec.shape, tensor_spec.dtype)
+            stack_trace = node.meta.get("stack_trace")
+            fqn = _get_module_hierarchy(node)
+            for j in range(start, end + 1):
+                if memory_timeline[j] is None:
+                    memory_timeline[j] = MemoryTimeline()
+                memory_timeline[j].allocations.append(
```

Comment:
can we keep the lifetime in the `Allocation` till we want to generate the trace so it might be more compressed?

Reply:
Sure. This just makes code more readable, but if this results in other perf issues, then your comment is definitely valid.